src/osd/PG.cc

```diff
@@ -1429,23 +1429,19 @@ bool PG::choose_acting(pg_shard_t &auth_log_shard_id,
     }
   }
 
-  // We go incomplete if below min_size for ec_pools since backfill
-  // does not currently maintain rollbackability
-  // Otherwise, we will go "peered", but not "active"
-  if (num_want_acting < pool.info.min_size &&
-      (pool.info.is_erasure() ||
-       !cct->_conf->osd_allow_recovery_below_min_size)) {
+  /* Check whether we have enough acting shards to later perform recovery */
+  auto recoverable_predicate = get_pgbackend()->get_is_recoverable_predicate();
+  if (!(*recoverable_predicate)(have)) {
     want_acting.clear();
-    dout(10) << "choose_acting failed, below min size" << dendl;
+    dout(10) << "choose_acting failed, not recoverable" << dendl;
     return false;
   }
 
-  /* Check whether we have enough acting shards to later perform recovery */
-  boost::scoped_ptr<IsPGRecoverablePredicate> recoverable_predicate(
-    get_pgbackend()->get_is_recoverable_predicate());
-  if (!(*recoverable_predicate)(have)) {
+  if (!cct->_conf->osd_allow_recovery_below_min_size && 
+      pool.info.is_replicated() &&
```

Comment:
While osd_allow_recovery_below_min_size definitely exists for debugging, there's no reason to restrict it to replicated pools.

Reply:
Sorry for my late reply.  I tried to keep the code consistent with the previous. and I found that `osd_allow_recovery_below_min_size` is only used in here. maybe we could drop this option?

Reply:
No, I mean that's an option mean to be used for testing and debugging. Replicated pools have various features that let them simulate an EC pool to deal with some of the ways in which peering is different between the two.

In this case, replicated pools have always been able to do recovery when below min_size, but we needed to be able to disable it so the behavior matches EC pools. Now that this PR is allowing recovery from any size, we may want this option to work on both pool types.

Although you're right; it becomes less useful once all pools can just recover whatever their size. Hrm.